### PR TITLE
Programmatically add group version and kind information for kcp compatibility 

### PR DESCRIPTION
**Pull request by shangma**
Resolve issuse: [#8](https://github.com/rancher/support-bundle-kit/issues/8) from support-bundle-kit side

Changes:
    1. programmatically add gvk info to the generated yaml file
    2. replace empty string ("") to "null" for kcp compatibility
    3. replace json null with yaml "null"
    4. each yaml is a list of resources with kind of list and apiVersion of v1
    5. remove duplicate resources generation code in old style

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -45,6 +45,7 @@ replace (
 )
 
 require (
+	github.com/Jeffail/gabs/v2 v2.6.1
```
Comment (yasker) [inline go.mod:48]: Separate the `vendor` folder change with `go.mod` change into the first commit, so it won't break the build if uses `git bisect`.

File: pkg/manager/client/discovery.go [hunk 2]
```diff
@@ -1,10 +1,11 @@
 package client
 
 import (
-	"bytes"
 	"context"
-	"encoding/json"
 	"fmt"
+	"github.com/Jeffail/gabs/v2"
```
Comment (yasker) [inline pkg/manager/client/discovery.go]: Move `github.com` import into the second section of the import, along with logrus.

File: pkg/manager/client/discovery.go [hunk 3]
```diff
@@ -30,26 +31,46 @@ func NewDiscoveryClient(ctx context.Context, config *rest.Config) (*DiscoveryCli
 	}, nil
 }
 
-func toObj(b []byte) interface{} {
-	var objMap map[string]interface{}
+func toObj(b []byte, groupVersion string, kind string) interface{} {
```
Comment (yasker) [inline pkg/manager/client/discovery.go]: If the function can fail, we should return the error code here as well.

File: pkg/manager/client/discovery.go [hunk 4]
```diff
@@ -93,7 +114,7 @@ func (dc *DiscoveryClient) ResourcesForNamespace(namespace string) map[string]in
 			b, err := result.Raw()
 
 			if err == nil {
-				obj := toObj(b)
+				obj := toObj(b, gv.String(), resource.Kind)
```
Comment (yasker) [inline pkg/manager/client/discovery.go:111]: Yeah, toObj should return both obj and error.

File: pkg/manager/cluster.go [hunk 5]
```diff
@@ -81,20 +81,11 @@ func (c *Cluster) GenerateClusterBundle(bundleDir string) (string, error) {
 func (c *Cluster) generateSupportBundleYAMLs(yamlsDir string, errLog io.Writer) {
 	// Cluster scope
 	globalDir := filepath.Join(yamlsDir, "cluster")
-	c.generateKubernetesClusterYAMLs(globalDir, errLog)
 	c.generateDiscoveredClusterYAMLs(globalDir, errLog)
 
-	// Namespaced scope: k8s resources
+	// Namespaced scope: all resources
```
Comment (yasker) [inline pkg/manager/cluster.go:86]: Separate this part into another commit, since it's not related to group version etc change.

File: pkg/manager/client/discovery.go [hunk 6]
```diff
@@ -31,40 +30,35 @@ func NewDiscoveryClient(ctx context.Context, config *rest.Config) (*DiscoveryCli
...
-	re, err := regexp.Compile(`""`)
-	if err != nil {
-		logrus.Errorf("Unable to create regex.")
-		return nil
-	}
+func toObj(b []byte, namespace, groupVersion, kind string) interface{} {
 
-	replacement := `"null"`
-	afterReplace := re.ReplaceAllLiteralString(string(b), replacement)
+	replaceString := strings.ReplaceAll(string(b), `"creationTimestamp":null`, `"creationTimestamp":"null"`)
```
Comment (yasker) [inline pkg/manager/client/discovery.go:35]: Hmm? How can `creationTimestamp` is null for an existing object? And why is it treated as a special case?

File: pkg/manager/client/discovery.go [hunk 7]
```diff
@@ -31,40 +30,35 @@ func NewDiscoveryClient(ctx context.Context, config *rest.Config) (*DiscoveryCli
...
 	if err != nil {
-		logrus.Errorf("Unable to parse json.")
+		logrus.Errorf("Unable to parse json: %s, %s, %s", namespace, groupVersion, kind)
 		return nil
 	}
-
 	// the yaml contains a list of resources
-	_, err = jsonParsed.SetP("List", "kind")
-	_, err = jsonParsed.SetP("apps/v1", "apiVersion")
+	if _, err = jsonParsed.SetP("List", "kind"); err != nil {
```
Comment (yasker) [inline pkg/manager/client/discovery.go:45]: It seems this commit has similar functionality as the first one. We should merge them if it's the case.

File: vendor/modules.txt [hunk 8]
```diff
@@ -1,3 +1,6 @@
+# github.com/Jeffail/gabs/v2 v2.6.1
+## explicit
+github.com/Jeffail/gabs/v2
```
Comment (yasker) [inline vendor/modules.txt:3]: Should be in the first and separated commit as mentioned in the first commit's comment.

File: pkg/manager/client/discovery.go [hunk 9]
```diff
@@ -30,15 +30,15 @@ func NewDiscoveryClient(ctx context.Context, config *rest.Config) (*DiscoveryCli
...
+func toObj(b []byte, groupVersion, kind string) interface{} {
 
 	replaceString := strings.ReplaceAll(string(b), `"creationTimestamp":null`, `"creationTimestamp":"null"`)
 	replaceString = strings.ReplaceAll(replaceString, `\"creationTimestamp\":null`, `\"creationTimestamp\":\"null\"`)
 
 	finalString := strings.ReplaceAll(replaceString, `""`, `"null"`)
 	jsonParsed, err := gabs.ParseJSON([]byte(finalString))
 	if err != nil {
-		logrus.Errorf("Unable to parse json: %s, %s, %s", namespace, groupVersion, kind)
+		logrus.Errorf("Unable to parse json: %s, %s", groupVersion, kind)
```
Comment (yasker) [inline pkg/manager/client/discovery.go:41]: This part needs to be cleaned up/merged with other commits. This doesn't belong to the vendor change.

**Review by yasker — CHANGES_REQUESTED** [review]

File: pkg/manager/client/discovery.go [hunk 10]
```diff
@@ -30,35 +30,50 @@ func NewDiscoveryClient(ctx context.Context, config *rest.Config) (*DiscoveryCli
...
-	r := bytes.NewReader(b)
-	d := json.NewDecoder(r)
+	replaceString := strings.ReplaceAll(string(b), `"creationTimestamp":null`, `"creationTimestamp":"null"`)
+	replaceString = strings.ReplaceAll(replaceString, `\"creationTimestamp\":null`, `\"creationTimestamp\":\"null\"`)
 
-	if err := d.Decode(&objMap); err != nil {
-		logrus.Errorf("Unable to decode json.")
-		return nil
+	finalString := strings.ReplaceAll(replaceString, `""`, `"null"`)
+	jsonParsed, err := gabs.ParseJSON([]byte(finalString))
```
Comment (yasker) [inline pkg/manager/client/discovery.go:39]: I just noticed. Why it's JSON format instead of yaml?

Comment (shangma) [reply pkg/manager/client/discovery.go:39]: The result is a byte slice which gets decode to json for further manipulation and then gets serialized  as yaml file here https://github.com/shangma/support-bundle-kit/blob/a422256672ef941438c37c1f347f9e7db38dd363/pkg/manager/cluster.go#L145

File: pkg/manager/client/discovery.go (hunk 6 above)
Comment (shangma) [reply pkg/manager/client/discovery.go:35]: The special case is due to the fact that json requires using null (without double quote) to represent a null field, but yaml needs a string with proper quotes. Let me dig into why the returned timestamp is a null field. @yasker

File: pkg/manager/client/discovery.go (hunk 10 above)
Comment (yasker) [reply pkg/manager/client/discovery.go:39]: Need to add comments in https://github.com/shangma/support-bundle-kit/blob/a422256672ef941438c37c1f347f9e7db38dd363/pkg/manager/cluster.go#L145 regarding why cannot we directly encode into YAML. I assume it's because of some of the options?

File: pkg/manager/client/discovery.go (hunk 6 above)
Comment (yasker) [reply pkg/manager/client/discovery.go:35]: I meant, `null vs "null"` should apply to all fields, not only `creationTimestamp`.

File: pkg/manager/client/discovery.go (hunk 10 above)
Comment (shangma) [reply pkg/manager/client/discovery.go:43]: Just did. @yasker Looks like no matter even yaml is setting to true in the option, still only a subset of it which is json compatible can be generated.

File: pkg/manager/client/discovery.go (hunk 6 above)
Comment (shangma) [reply pkg/manager/client/discovery.go]: @yasker  absolutely make much sense. Just update it to match all invalid null in the format with regexp.

**Review by yasker — APPROVED** [review]
